### PR TITLE
Implement thread safety during instantiation of TelemetryService

### PR DESCRIPTION
**Pull request by sfc-gh-yqiu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes [SNOW-895151: Non Thread Safe Singleton Telemetry Service in Python](https://github.com/snowflakedb/snowflake-connector-python/issues/1700)

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Adding a lock guarding the instantiation of `TelemetryService` prevents multiple threads from entering `__init__` and violating the singleton constraint.

### CODE REVIEW

**Comment by sfc-gh-yqiu** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by sfc-gh-aling — COMMENTED** [review]
the changes LGTM.

it looks like there're some flaky test failures unrelated to your changes, usually retry running the test would help.

also please update the changelog

File: DESCRIPTION.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ Source code is also available at: https://github.com/snowflakedb/snowflake-conne
 
 # Release Notes
 
+- v3.2.1(TBD)
+
+  - Added thread safety when instantiating multiple connections
```
Comment (sfc-gh-aling) [inline DESCRIPTION.md:13]: nit, can be a bit more precise:
```suggestion
  - Added thread safety in telemetry when instantiating multiple connections concurrently.
```

**Review by sfc-gh-aling — APPROVED** [review]
LGTM!

**Pull request merged**